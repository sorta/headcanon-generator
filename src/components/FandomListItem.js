import React, { Component } from 'react';
import PropTypes from 'prop-types';

class FandomList extends Component {
  static propTypes = {
    fandom: PropTypes.shape({ name: PropTypes.string }),
    selectFandom: PropTypes.func,
    updateFandom: PropTypes.func,
    deleteFandom: PropTypes.func,
    isManaging: PropTypes.func,
    selectedFandomKey: PropTypes.string,
    index: PropTypes.string,
  };
  static defaultProps = {
    fandom: { name: '' },
    selectFandom: () => {},
    updateFandom: () => {},
    deleteFandom: () => {},
    isManaging: () => false,
    selectedFandomKey: '',
    index: '',
  };

  handleChange = (event) => {
    const updatedFandom = {
      ...this.props.fandom,
      [event.currentTarget.name]: event.currentTarget.value,
    };

    this.props.updateFandom(this.props.index, updatedFandom);
  };

  render() {
    const isSelected = this.props.index === this.props.selectedFandomKey;
    let itemContent = (
      <div className="control-row">
        <span className="control-stretch">{this.props.fandom.name}</span>
        <button onClick={() => this.props.selectFandom(isSelected ? '' : this.props.index)}>
          {isSelected ? '<<' : '>>'}
        </button>
      </div>
    );

    if (this.props.isManaging()) {
      itemContent = (
        <div className="control-row">
          <input name="name" type="text" placeholder="Fandom Name" className="control-stretch"
            value={this.props.fandom.name} onChange={this.handleChange}
          />
          <button onClick={() => this.props.deleteFandom(this.props.index)}>-</button>
          <button onClick={() => this.props.selectFandom(isSelected ? '' : this.props.index)}>
            {isSelected ? '<<' : '>>'}
          </button>
        </div>
      );
    }

    return (
      <li key={this.props.index} className="FandomListItem list-item">
        {itemContent}
      </li>
    );
  }
}

export default FandomList;
