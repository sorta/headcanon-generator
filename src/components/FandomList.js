import React, { Component } from 'react';
import PropTypes from 'prop-types';

import FandomListItem from './FandomListItem'

class FandomList extends Component {
  static propTypes = {
    fandoms: PropTypes.object,
    selectFandom: PropTypes.func,
    updateFandom: PropTypes.func,
    deleteFandom: PropTypes.func,
    isManaging: PropTypes.func,
    selectedFandomKey: PropTypes.string,
  };
  static defaultProps = {
    fandoms: {},
    selectFandom: () => {},
    updateFandom: () => {},
    deleteFandom: () => {},
    isManaging: () => false,
    selectedFandomKey: '',
  };

  render() {
    if (Object.keys(this.props.fandoms).length === 0) {
      return null;
    }

    return (
      <ul className="options-list FandomList">
        {Object.keys(this.props.fandoms).map(key => {
          if (!{}.hasOwnProperty.call(this.props.fandoms, key)) {
            return null;
          }

          return <FandomListItem
            key={key}
            index={key}
            fandom={this.props.fandoms[key]}
            selectFandom={this.props.selectFandom}
            updateFandom={this.props.updateFandom}
            deleteFandom={this.props.deleteFandom}
            isManaging={this.props.isManaging}
            selectedFandomKey={this.props.selectedFandomKey}
          />
        })}
      </ul>
    );
  }
}

export default FandomList;
