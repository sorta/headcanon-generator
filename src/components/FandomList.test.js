import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import FandomList from './FandomList';
import FandomListItem from './FandomListItem';

const props = {
  fandoms: {
    f_test_1: { name: 'Test 1' },
    f_test_2: { name: 'Test 2' },
    f_test_3: { name: 'Test 3' },
  },
  fandomAvailability: {},
  updateFandom: jest.fn(),
  deleteFandom: jest.fn(),
  setAvailability: jest.fn(),
  isManaging: () => false,
  selectFandom: jest.fn(),
  selectedFandomKey: '',
};

// Tests for regular mode
describe("FandomList", () => {

  it('renders and matches our snapshot', () => {
    const snapshot = renderer.create(
      <FandomList { ...props } />
    );
    const tree = snapshot.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('has a FandomList ul element', () => {
    const component = shallow(<FandomList { ...props } />);
    expect(component.find('ul.FandomList.options-list').exists()).toBe(true);
  });

  it('has right number of FandomListItems', () => {
    const component = shallow(<FandomList { ...props } />);
    expect(component.find(FandomListItem)).toHaveLength(Object.keys(props.fandoms).length);
  });
});
