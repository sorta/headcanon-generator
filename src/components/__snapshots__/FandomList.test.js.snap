// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FandomList renders and matches our snapshot 1`] = `
<ul
  className="options-list FandomList"
>
  <li
    className="FandomListItem list-item control-row"
  >
    <input
      checked={true}
      className="cb-fake"
      id="isAvailable_fand_f_test_1"
      name="isAvailable"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="FandomListItem-text control-stretch"
      htmlFor="isAvailable_fand_f_test_1"
    >
      Test 1
    </label>
    <button
      onClick={[Function]}
    >
      &gt;&gt;
    </button>
  </li>
  <li
    className="FandomListItem list-item control-row"
  >
    <input
      checked={true}
      className="cb-fake"
      id="isAvailable_fand_f_test_2"
      name="isAvailable"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="FandomListItem-text control-stretch"
      htmlFor="isAvailable_fand_f_test_2"
    >
      Test 2
    </label>
    <button
      onClick={[Function]}
    >
      &gt;&gt;
    </button>
  </li>
  <li
    className="FandomListItem list-item control-row"
  >
    <input
      checked={true}
      className="cb-fake"
      id="isAvailable_fand_f_test_3"
      name="isAvailable"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="FandomListItem-text control-stretch"
      htmlFor="isAvailable_fand_f_test_3"
    >
      Test 3
    </label>
    <button
      onClick={[Function]}
    >
      &gt;&gt;
    </button>
  </li>
</ul>
`;
